File: app/src/main/java/com/appzonegroup/app/fasttrack/utility/TrackGPS.java
package com.appzonegroup.app.fasttrack.utility;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.os.Bundle;
import android.provider.Settings;
import android.util.Log;
import android.widget.Toast;

import com.creditclub.core.data.prefs.LocalStorage;
import com.creditclub.core.util.ContextExtensionsKt;

import java.util.Locale;

import static android.content.Context.LOCATION_SERVICE;

public class TrackGPS implements LocationListener {

    private final Context mContext;

    boolean checkGPS = false, checkNetwork = false, canGetLocation = false;

    Location loc;

    double latitude, longitude;


    LocationManager locationManager;

    public TrackGPS(Context mContext) {
        this.mContext = mContext.getApplicationContext();
        getLocation();
    }


    private Location getLocation() {

        try {
            locationManager = (LocationManager) mContext.getSystemService(LOCATION_SERVICE);

            // getting GPS status
            checkGPS = locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER);

            // getting network status
            checkNetwork = locationManager
                    .isProviderEnabled(LocationManager.NETWORK_PROVIDER);

            long MIN_DISTANCE_CHANGE_FOR_UPDATES = 10;
            long MIN_TIME_BW_UPDATES = 3000;
            if (!checkGPS && !checkNetwork) {
                Toast.makeText(mContext, "No Service Provider Available", Toast.LENGTH_SHORT).show();
            } else {
                this.canGetLocation = true;
                // First get location from Network Provider
                if (checkNetwork) {
                    // Toast.makeText(mContext, "Network", Toast.LENGTH_SHORT).show();

                    try {
                        locationManager.requestLocationUpdates(
                                LocationManager.NETWORK_PROVIDER,
                                MIN_TIME_BW_UPDATES,
                                MIN_DISTANCE_CHANGE_FOR_UPDATES, this);
                        Log.d("Network", "Network");
                        if (locationManager != null) {
                            loc = locationManager
                                    .getLastKnownLocation(LocationManager.NETWORK_PROVIDER);

                        }

                        if (loc != null) {
                            latitude = loc.getLatitude();
                            longitude = loc.getLongitude();
                        }
                    } catch (SecurityException e) {

                    }
                }
            }
            // if GPS Enabled get lat/long using GPS Services
            if (checkGPS) {
                //Toast.makeText(mContext, "GPS", Toast.LENGTH_SHORT).show();
                if (loc == null) {
                    try {
                        locationManager.requestLocationUpdates(
                                LocationManager.GPS_PROVIDER,
                                MIN_TIME_BW_UPDATES,
                                MIN_DISTANCE_CHANGE_FOR_UPDATES, this);
                        Log.d("GPS Enabled", "GPS Enabled");
                        if (locationManager != null) {
                            loc = locationManager
                                    .getLastKnownLocation(LocationManager.GPS_PROVIDER);
                            if (loc != null) {
                                latitude = loc.getLatitude();
                                longitude = loc.getLongitude();

                                LocalStorage localStorage = ContextExtensionsKt.getLocalStorage(mContext);
                                localStorage.setLastKnownLocation(String.format(Locale.getDefault(), "%.2f;%.2f", latitude, longitude));
                            }
                        }
                    } catch (SecurityException e) {

                    }
                }
            }

        } catch (Exception e) {
            e.printStackTrace();
        }

        return loc;
    }

    public double getLongitude() {
        if (loc != null) {
            longitude = loc.getLongitude();
        }
        return longitude;
    }

    public double getLatitude() {
        if (loc != null) {
            latitude = loc.getLatitude();
        }
        return latitude;
    }

    public String getGeolocationString() {
        if (loc != null) {
            return String.format(Locale.getDefault(), "%.2f;%.2f", latitude, longitude);
        }
        return ContextExtensionsKt.getLocalStorage(mContext).getLastKnownLocation();
    }

    public boolean canGetLocation() {
        return this.canGetLocation;
    }

    public void showSettingsAlert() {
        AlertDialog.Builder alertDialog = new AlertDialog.Builder(mContext);


        alertDialog.setTitle("GPS Not Enabled");

        alertDialog.setMessage("Do you wants to turn On GPS");


        alertDialog.setPositiveButton("Yes", new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int which) {
                Intent intent = new Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS);
                mContext.startActivity(intent);
            }
        });


        alertDialog.setNegativeButton("No", new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int which) {
                dialog.cancel();
            }
        });


        alertDialog.show();
    }

    @Override
    public void onLocationChanged(Location location) {

    }

    @Override
    public void onStatusChanged(String provider, int status, Bundle extras) {

    }

    @Override
    public void onProviderEnabled(String provider) {

    }

    @Override
    public void onProviderDisabled(String provider) {

    }
}

File: core/src/main/java/com/creditclub/core/util/TrackGPS.kt
package com.creditclub.core.util

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.location.Location
import android.location.LocationListener
import android.location.LocationManager
import android.os.Bundle
import android.os.Looper
import android.util.Log
import android.widget.Toast
import androidx.core.app.ActivityCompat
import com.creditclub.core.data.prefs.LocalStorage
import org.koin.core.KoinComponent
import org.koin.core.inject

class TrackGPS(private val mContext: Context) : LocationListener, KoinComponent {
    private val localStorage: LocalStorage by inject()

    private var checkGPS = false
    private var checkNetwork = false
    private var canGetLocation = false

    private var loc: Location? = null

    var latitude: Double = 0.toDouble()
        get() {
            if (loc != null) {
                field = loc!!.latitude
            }
            return field
        }

    var longitude: Double = 0.toDouble()
        get() {
            if (loc != null) {
                field = loc!!.longitude
            }
            return field
        }

    val geolocationString: String?
        get() = if (loc == null) localStorage.lastKnownLocation
        else "$latitude;$longitude"

    private var locationManager: LocationManager? = null

    override fun onLocationChanged(location: Location) {
        loc = location
        localStorage.lastKnownLocation = "${location.latitude};${location.longitude}"
    }

    override fun onStatusChanged(provider: String, status: Int, extras: Bundle) {

    }

    override fun onProviderEnabled(provider: String) {

    }

    override fun onProviderDisabled(provider: String) {

    }

    init {
        safeRun {
            locationManager =
                mContext.getSystemService(Context.LOCATION_SERVICE) as LocationManager
            checkGPS = locationManager!!.isProviderEnabled(LocationManager.GPS_PROVIDER)
            checkNetwork = locationManager!!.isProviderEnabled(LocationManager.NETWORK_PROVIDER)

            if (!checkGPS && !checkNetwork) {
                Toast.makeText(mContext, "No Service Provider Available", Toast.LENGTH_SHORT)
                    .show()
            } else {
                this.canGetLocation = true
                if (checkNetwork) {
                    try {
                        locationManager?.requestLocationUpdates(
                            LocationManager.NETWORK_PROVIDER,
                            MIN_TIME_BW_UPDATES,
                            MIN_DISTANCE_CHANGE_FOR_UPDATES.toFloat(), this
                        )
                        Log.d("Network", "Network")
                        loc =
                            locationManager?.getLastKnownLocation(LocationManager.NETWORK_PROVIDER)
                                ?: loc

                        if (loc != null) {
                            latitude = loc!!.latitude
                            longitude = loc!!.longitude
                            localStorage.lastKnownLocation =
                                "${loc?.latitude};${loc?.longitude}"
                        }
                    } catch (e: SecurityException) {

                    }
                }
            }

            if (checkGPS) {
                if (loc == null) {
                    Looper.myLooper() ?: Looper.prepare()
                    if (ActivityCompat.checkSelfPermission(
                            mContext,
                            Manifest.permission.ACCESS_FINE_LOCATION
                        ) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(
                            mContext,
                            Manifest.permission.ACCESS_COARSE_LOCATION
                        ) != PackageManager.PERMISSION_GRANTED
                    ) {
                        return@safeRun
                    }
                    locationManager!!.requestLocationUpdates(
                        LocationManager.GPS_PROVIDER,
                        MIN_TIME_BW_UPDATES,
                        MIN_DISTANCE_CHANGE_FOR_UPDATES.toFloat(), this
                    )
                    Log.d("GPS Enabled", "GPS Enabled")
                    if (locationManager != null) {
                        loc = locationManager!!
                            .getLastKnownLocation(LocationManager.GPS_PROVIDER)
                        if (loc != null) {
                            latitude = loc!!.latitude
                            longitude = loc!!.longitude
                        }
                    }
                }
            }
        }
    }

    companion object {
        const val MIN_DISTANCE_CHANGE_FOR_UPDATES: Long = 10
        const val MIN_TIME_BW_UPDATES: Long = 3000
    }
}
